File: Day063_2025/Q063_GFG.py
import bisect

class Solution:
    def lis(self, arr):
        n = len(arr)
        ans = [arr[0]]

        for i in range(1, n):
            if arr[i] > ans[-1]:
                ans.append(arr[i])
            else:
                idx = bisect.bisect_left(ans, arr[i])
                ans[idx] = arr[i]

        return len(ans)

File: Day063_2025/Q063_GFG.java
import java.util.*;

class Q063_GFG {
    public static int lis(int[] arr) {
        int n = arr.length;
        ArrayList<Integer> ans = new ArrayList<>();
        ans.add(arr[0]);

        for (int i = 1; i < n; i++) {
            if (arr[i] > ans.get(ans.size() - 1)) {
                ans.add(arr[i]);
            } else {
                int low = Collections.binarySearch(ans, arr[i]);
                if (low < 0) low = -(low + 1);
                ans.set(low, arr[i]);
            }
        }
        return ans.size();
    }
}
